src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+    ///     - https://github.com/dotnet/roslyn/blob/main/docs/compilers/Deterministic%20Inputs.md
+    /// 
+    /// Options which can cause compilation failure, but doesn't impact the result of a successful
+    /// compilation should be included. That is because it is interesting to describe error states
+    /// not just success states. Think about caching build failures as well as build successes.
+    ///
+    /// When an option is omitted, say if there is no value for a public crypto key, we should emit
+    /// the property with a null value vs. omitting the property. Either approach would produce 
+    /// correct results the preference is to be declarative that an option is omitted.
+    /// </summary>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFilePath(
+            JsonWriter writer,
+            string propertyName,
+            string? filePath,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
+            }
+            else if (filePath is not null)
+            {
+                filePath = PathUtilities.NormalizePathPrefix(filePath, pathMap);
+            }
+
+            writer.Write(propertyName, filePath);
+        }
+
+        internal static string EncodeByteArrayValue(ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            return builder.ToStringAndFree();
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected static void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected static void WriteVersion(JsonWriter writer, string key, Version version)
+        {
+            writer.WriteKey(key);
+            writer.WriteObjectStart();
+            writer.Write("major", version.Major);
+            writer.Write("minor", version.Minor);
+            writer.Write("build", version.Build);
+            writer.Write("revision", version.Revision);
+            writer.WriteObjectEnd();
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
```

Comment:
Do we have a test that shows what this outputs?

Reply:
Hmm, good catch. Thought I had added that but apparently missed that case.